mono msbuild fails on Linux/Mac when MsCorEE cannot be found

From [this log](https://travis-ci.org/AArnott/libgit2sharp/jobs/226030013#L2158)

```
/home/travis/.nuget/packages/nerdbank.gitversioning/1.6.30/build/Nerdbank.GitVersioning.targets(128,5): error MSB4018: This is an unhandled exception from a task -- PLEASE OPEN A BUG AGAINST THE TASK OWNER. [/home/travis/build/AArnott/libgit2sharp/CodeGenerationAttributes/CodeGenerationAttributes.csproj]
/home/travis/.nuget/packages/nerdbank.gitversioning/1.6.30/build/Nerdbank.GitVersioning.targets(128,5): error MSB4018: System.DllNotFoundException: MSCorEE [/home/travis/build/AArnott/libgit2sharp/CodeGenerationAttributes/CodeGenerationAttributes.csproj]
/home/travis/.nuget/packages/nerdbank.gitversioning/1.6.30/build/Nerdbank.GitVersioning.targets(128,5): error MSB4018:   at (wrapper managed-to-native) PInvoke.MSCorEE:StrongNameGetPublicKey (string,byte*,int,byte*&,int&) [/home/travis/build/AArnott/libgit2sharp/CodeGenerationAttributes/CodeGenerationAttributes.csproj]
/home/travis/.nuget/packages/nerdbank.gitversioning/1.6.30/build/Nerdbank.GitVersioning.targets(128,5): error MSB4018:   at PInvoke.MSCorEE.StrongNameGetPublicKey (System.Byte[] keyBlob) [0x00017] in <a4514f63a52a405fac466fdf826c8ba3>:0  [/home/travis/build/AArnott/libgit2sharp/CodeGenerationAttributes/CodeGenerationAttributes.csproj]
/home/travis/.nuget/packages/nerdbank.gitversioning/1.6.30/build/Nerdbank.GitVersioning.targets(128,5): error MSB4018:   at Nerdbank.GitVersioning.Tasks.AssemblyVersionInfo.ReadKeyInfo (System.String& publicKey, System.String& publicKeyToken) [0x0004b] in <819bf1bdc5464561afd45b3c0ae000c1>:0  [/home/travis/build/AArnott/libgit2sharp/CodeGenerationAttributes/CodeGenerationAttributes.csproj]
/home/travis/.nuget/packages/nerdbank.gitversioning/1.6.30/build/Nerdbank.GitVersioning.targets(128,5): error MSB4018:   at Nerdbank.GitVersioning.Tasks.AssemblyVersionInfo.CreateThisAssemblyClass () [0x0003e] in <819bf1bdc5464561afd45b3c0ae000c1>:0  [/home/travis/build/AArnott/libgit2sharp/CodeGenerationAttributes/CodeGenerationAttributes.csproj]
/home/travis/.nuget/packages/nerdbank.gitversioning/1.6.30/build/Nerdbank.GitVersioning.targets(128,5): error MSB4018:   at Nerdbank.GitVersioning.Tasks.AssemblyVersionInfo.Execute () [0x00060] in <819bf1bdc5464561afd45b3c0ae000c1>:0  [/home/travis/build/AArnott/libgit2sharp/CodeGenerationAttributes/CodeGenerationAttributes.csproj]
/home/travis/.nuget/packages/nerdbank.gitversioning/1.6.30/build/Nerdbank.GitVersioning.targets(128,5): error MSB4018:   at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute () [0x0002a] in <07a5e954af354541b797e7afb63dd350>:0  [/home/travis/build/AArnott/libgit2sharp/CodeGenerationAttributes/CodeGenerationAttributes.csproj]
/home/travis/.nuget/packages/nerdbank.gitversioning/1.6.30/build/Nerdbank.GitVersioning.targets(128,5): error MSB4018:   at Microsoft.Build.BackEnd.TaskBuilder+<ExecuteInstantiatedTask>d__26.MoveNext () [0x002bf] in <07a5e954af354541b797e7afb63dd350>:0  [/home/travis/build/AArnott/libgit2sharp/CodeGenerationAttributes/CodeGenerationAttributes.csproj]
```

--- a/src/Nerdbank.GitVersioning.Tasks/AssemblyVersionInfo.cs
+++ b/src/Nerdbank.GitVersioning.Tasks/AssemblyVersionInfo.cs
@@ -11,9 +11,7 @@
     using System.Text;
     using Microsoft.Build.Framework;
     using Microsoft.Build.Utilities;
-#if NET45
-    using PInvoke;
-#endif
+
     public class AssemblyVersionInfo : Task
     {
 #if NET45
@@ -121,7 +119,7 @@
 
             // Determine information about the public key used in the assembly name.
             string publicKey, publicKeyToken;
-            this.ReadKeyInfo(out publicKey, out publicKeyToken);
+            bool hasKeyInfo = this.TryReadKeyInfo(out publicKey, out publicKeyToken);
 
             // Define the constants.
             thisAssembly.Members.AddRange(CreateFields(new Dictionary<string, string>
@@ -130,58 +128,25 @@
                     { "AssemblyFileVersion", this.AssemblyFileVersion },
                     { "AssemblyInformationalVersion", this.AssemblyInformationalVersion },
                     { "AssemblyName", this.AssemblyName },
-                    { "PublicKey", publicKey },
-                    { "PublicKeyToken", publicKeyToken },
                     { "AssemblyTitle", this.AssemblyTitle },
                     { "AssemblyProduct", this.AssemblyProduct },
                     { "AssemblyCopyright", this.AssemblyCopyright },
                     { "AssemblyCompany", this.AssemblyCompany },
                     { "AssemblyConfiguration", this.AssemblyConfiguration }
                 }).ToArray());
+            if (hasKeyInfo)
+            {
+                thisAssembly.Members.AddRange(CreateFields(new Dictionary<string, string>
+                {
+                    { "PublicKey", publicKey },
+                    { "PublicKeyToken", publicKeyToken },
+                }).ToArray());
+            }
 
             // These properties should be defined even if they are empty.
             thisAssembly.Members.Add(CreateField("RootNamespace", this.RootNamespace));
 
             return thisAssembly;
-        }
-
-       private void ReadKeyInfo(out string publicKey, out string publicKeyToken)
-        {
-            byte[] publicKeyBytes = null;
-            if (!string.IsNullOrEmpty(this.AssemblyOriginatorKeyFile) && File.Exists(this.AssemblyOriginatorKeyFile))
-            {
-                if (Path.GetExtension(this.AssemblyOriginatorKeyFile).Equals(".snk", StringComparison.OrdinalIgnoreCase))
-                {
-                    byte[] keyBytes = File.ReadAllBytes(this.AssemblyOriginatorKeyFile);
-                    bool publicKeyOnly = keyBytes[0] != 0x07;
-                    publicKeyBytes = publicKeyOnly ? keyBytes : MSCorEE.StrongNameGetPublicKey(keyBytes);
-                }
-            }
-            else if (!string.IsNullOrEmpty(this.AssemblyKeyContainerName))
-            {
-                publicKeyBytes = MSCorEE.StrongNameGetPublicKey(this.AssemblyKeyContainerName);
-            }
-
-            if (publicKeyBytes != null && publicKeyBytes.Length > 0) // If .NET 2.0 isn't installed, we get byte[0] back.
-            {
-                publicKey = ToHex(publicKeyBytes);
-                publicKeyToken = ToHex(MSCorEE.StrongNameTokenFromPublicKey(publicKeyBytes));
-            }
-            else
-            {
-                if (publicKeyBytes != null)
-                {
-                    this.Log.LogWarning("Unable to emit public key fields in ThisAssembly class because .NET 2.0 isn't installed.");
-                }
-
-                publicKey = null;
-                publicKeyToken = null;
-            }
-        }
-
-        private static string ToHex(byte[] data)
-        {
-            return BitConverter.ToString(data).Replace("-", "").ToLowerInvariant();
         }
 
         private IEnumerable<CodeAttributeDeclaration> CreateAssemblyAttributes()
@@ -276,9 +241,9 @@
         {
             this.generator.StartThisAssemblyClass();
 
-            ////// Determine information about the public key used in the assembly name.
-            ////string publicKey, publicKeyToken;
-            ////this.ReadKeyInfo(out publicKey, out publicKeyToken);
+            // Determine information about the public key used in the assembly name.
+            string publicKey, publicKeyToken;
+            bool hasKeyInfo = this.TryReadKeyInfo(out publicKey, out publicKeyToken);
 
             // Define the constants.
             var fields = new Dictionary<string, string>
@@ -287,14 +252,18 @@
                     { "AssemblyFileVersion", this.AssemblyFileVersion },
                     { "AssemblyInformationalVersion", this.AssemblyInformationalVersion },
                     { "AssemblyName", this.AssemblyName },
-                    ////{ "PublicKey", publicKey },
-                    ////{ "PublicKeyToken", publicKeyToken },
                     { "AssemblyTitle", this.AssemblyTitle },
                     { "AssemblyProduct", this.AssemblyProduct },
                     { "AssemblyCopyright", this.AssemblyCopyright },
                     { "AssemblyCompany", this.AssemblyCompany },
                     { "AssemblyConfiguration", this.AssemblyConfiguration }
                 };
+            if (hasKeyInfo)
+            {
+                fields.Add("PublicKey", publicKey);
+                fields.Add("PublicKeyToken", publicKeyToken);
+            }
+
             foreach (var pair in fields)
             {
                 if (!string.IsNullOrEmpty(pair.Value))
@@ -389,5 +358,79 @@
             }
         }
 #endif
+
+        private static string ToHex(byte[] data)
+        {
+            return BitConverter.ToString(data).Replace("-", "").ToLowerInvariant();
+        }
+
+        /// <summary>
+        /// Gets the public key from a key container.
+        /// </summary>
+        /// <param name="containerName">The name of the container.</param>
+        /// <returns>The public key.</returns>
+        private static byte[] GetPublicKeyFromKeyContainer(string containerName)
+        {
+            throw new NotImplementedException();
+        }
+
+        private static byte[] GetPublicKeyFromKeyPair(byte[] keyPair)
+        {
+            byte[] publicKey;
+            if (CryptoBlobParser.TryGetPublicKeyFromPrivateKeyBlob(keyPair, out publicKey))
+            {
+                return publicKey;
+            }
+            else
+            {
+                throw new ArgumentException("Invalid keypair");
+            }
+        }
+
+        private bool TryReadKeyInfo(out string publicKey, out string publicKeyToken)
+        {
+            try
+            {
+                byte[] publicKeyBytes = null;
+                if (!string.IsNullOrEmpty(this.AssemblyOriginatorKeyFile) && File.Exists(this.AssemblyOriginatorKeyFile))
+                {
+                    if (Path.GetExtension(this.AssemblyOriginatorKeyFile).Equals(".snk", StringComparison.OrdinalIgnoreCase))
+                    {
+                        byte[] keyBytes = File.ReadAllBytes(this.AssemblyOriginatorKeyFile);
+                        bool publicKeyOnly = keyBytes[0] != 0x07;
+                        publicKeyBytes = publicKeyOnly ? keyBytes : GetPublicKeyFromKeyPair(keyBytes);
+                    }
+                }
+                else if (!string.IsNullOrEmpty(this.AssemblyKeyContainerName))
+                {
+                    publicKeyBytes = GetPublicKeyFromKeyContainer(this.AssemblyKeyContainerName);
+                }
+
+                if (publicKeyBytes != null && publicKeyBytes.Length > 0) // If .NET 2.0 isn't installed, we get byte[0] back.
+                {
+                    publicKey = ToHex(publicKeyBytes);
+                    publicKeyToken = ToHex(CryptoBlobParser.GetStrongNameTokenFromPublicKey(publicKeyBytes));
+                }
+                else
+                {
+                    if (publicKeyBytes != null)
+                    {
+                        this.Log.LogWarning("Unable to emit public key fields in ThisAssembly class because .NET 2.0 isn't installed.");
+                    }
+
+                    publicKey = null;
+                    publicKeyToken = null;
+                    return false;
+                }
+
+                return true;
+            }
+            catch (NotImplementedException)
+            {
+                publicKey = null;
+                publicKeyToken = null;
+                return false;
+            }
+        }
     }
 }

--- a/src/Nerdbank.GitVersioning.Tasks/CryptoBlobParser+BlobHeader.cs
+++ b/src/Nerdbank.GitVersioning.Tasks/CryptoBlobParser+BlobHeader.cs
@@ -1,0 +1,34 @@
+﻿namespace Nerdbank.GitVersioning.Tasks
+{
+    using System.Runtime.InteropServices;
+
+    /// <contents>
+    /// The <see cref="BlobHeader"/> struct.
+    /// </contents>
+    partial class CryptoBlobParser
+    {
+        [StructLayout(LayoutKind.Sequential)]
+        private struct BlobHeader
+        {
+            /// <summary>
+            /// Blob type.
+            /// </summary>
+            public byte Type;
+
+            /// <summary>
+            /// Blob format version
+            /// </summary>
+            public byte Version;
+
+            /// <summary>
+            /// Must be 0
+            /// </summary>
+            public ushort Reserved;
+
+            /// <summary>
+            /// Algorithm ID. Must be one of ALG_ID specified in wincrypto.h
+            /// </summary>
+            public uint AlgId;
+        }
+    }
+}

--- a/src/Nerdbank.GitVersioning.Tasks/CryptoBlobParser+RsaPubKey.cs
+++ b/src/Nerdbank.GitVersioning.Tasks/CryptoBlobParser+RsaPubKey.cs
@@ -1,0 +1,29 @@
+﻿namespace Nerdbank.GitVersioning.Tasks
+{
+    using System.Runtime.InteropServices;
+
+    partial class CryptoBlobParser
+    {
+        /// <summary>
+        /// RSAPUBKEY struct from wincrypt.h
+        /// </summary>
+        [StructLayout(LayoutKind.Sequential)]
+        private struct RsaPubKey
+        {
+            /// <summary>
+            /// Indicates RSA1 or RSA2
+            /// </summary>
+            public uint Magic;
+
+            /// <summary>
+            /// Number of bits in the modulus. Must be multiple of 8.
+            /// </summary>
+            public uint BitLen;
+
+            /// <summary>
+            /// The public exponent
+            /// </summary>
+            public uint PubExp;
+        }
+    }
+}

--- a/src/Nerdbank.GitVersioning.Tasks/CryptoBlobParser+SnPublicKeyBlob.cs
+++ b/src/Nerdbank.GitVersioning.Tasks/CryptoBlobParser+SnPublicKeyBlob.cs
@@ -1,0 +1,41 @@
+﻿namespace Nerdbank.GitVersioning.Tasks
+{
+    using System.Runtime.InteropServices;
+
+    /// <contents>
+    /// The <see cref="SnPublicKeyBlob"/> struct.
+    /// </contents>
+    partial class CryptoBlobParser
+    {
+        /// <summary>
+        /// The strong name public key blob binary format.
+        /// </summary>
+        /// <see cref="https://github.com/dotnet/coreclr/blob/32f0f9721afb584b4a14d69135bea7ddc129f755/src/strongname/inc/strongname.h#L29"/>
+        [StructLayout(LayoutKind.Sequential, Pack = 1)]
+        private unsafe struct SnPublicKeyBlob
+        {
+            /// <summary>
+            /// Signature algorithm ID
+            /// </summary>
+            public uint SigAlgId;
+
+            /// <summary>
+            /// Hash algorithm ID
+            /// </summary>
+            public uint HashAlgId;
+
+            /// <summary>
+            /// Size of public key data in bytes, not including the header
+            /// </summary>
+            public uint PublicKeySize;
+
+            /// <summary>
+            /// PublicKeySize bytes of public key data
+            /// </summary>
+            /// <remarks>
+            /// Note: PublicKey is variable sized
+            /// </remarks>
+            public fixed byte PublicKey[1];
+        }
+    }
+}

--- a/src/Nerdbank.GitVersioning.Tasks/CryptoBlobParser.cs
+++ b/src/Nerdbank.GitVersioning.Tasks/CryptoBlobParser.cs
@@ -1,0 +1,165 @@
+﻿// Liberally copied from (with slight modifications) https://github.com/dotnet/roslyn/blob/6181abfdf59da26da27f0dbedae2978df2f83768/src/Compilers/Core/Portable/StrongName/CryptoBlobParser.cs
+// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See https://github.com/dotnet/roslyn/blob/6181abfdf59da26da27f0dbedae2978df2f83768/License.txt for license information.
+
+namespace Nerdbank.GitVersioning.Tasks
+{
+    using System;
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Runtime.InteropServices;
+    using System.Security.Cryptography;
+    using System.Text;
+    using Validation;
+
+    internal static partial class CryptoBlobParser
+    {
+        /// <summary>
+        /// The size of a public key token, in bytes.
+        /// </summary>
+        private const int SN_SIZEOF_TOKEN = 8;
+
+        /// <summary>
+        /// The length of a SHA1 hash, in bytes.
+        /// </summary>
+        private const int SHA1_HASH_SIZE = 20;
+
+        private const UInt32 RSA1 = 0x31415352;
+
+        // From wincrypt.h
+        private const byte PublicKeyBlobId = 0x06;
+        private const byte PrivateKeyBlobId = 0x07;
+
+        // In wincrypt.h both public and private key blobs start with a
+        // PUBLICKEYSTRUC and RSAPUBKEY and then start the key data
+        private unsafe static readonly int s_offsetToKeyData = sizeof(BlobHeader) + sizeof(RsaPubKey);
+
+        /// <summary>
+        /// Derives the public key token from a full public key.
+        /// </summary>
+        /// <param name="publicKey">The public key.</param>
+        /// <returns>The public key token.</returns>
+        /// <remarks>
+        /// Heavily inspired by https://github.com/dotnet/coreclr/blob/2efbb9282c059eb9742ba5a59b8a1d52ac4dfa4c/src/strongname/api/strongname.cpp#L270
+        /// </remarks>
+        internal static unsafe byte[] GetStrongNameTokenFromPublicKey(byte[] publicKey)
+        {
+            Requires.NotNull(publicKey, nameof(publicKey));
+
+            byte[] strongNameToken = new byte[SN_SIZEOF_TOKEN];
+            fixed (byte* publicKeyPtr = publicKey)
+            {
+                //var publicKeyBlob = (SnPublicKeyBlob*)publicKeyPtr;
+                using (var sha1 = SHA1.Create())
+                {
+                    byte[] hash = sha1.ComputeHash(publicKey);
+                    int hashLenMinusTokenSize = SHA1_HASH_SIZE - SN_SIZEOF_TOKEN;
+
+                    // Take the last few bytes of the hash value for our token. (These are the
+                    // low order bytes from a network byte order point of view). Reverse the
+                    // order of these bytes in the output buffer to get host byte order.
+                    for (int i = 0; i < SN_SIZEOF_TOKEN; i++)
+                    {
+                        strongNameToken[SN_SIZEOF_TOKEN - (i + 1)] = hash[i + hashLenMinusTokenSize];
+                    }
+                }
+            }
+
+            return strongNameToken;
+        }
+
+        private static byte[] CreateSnPublicKeyBlob(BlobHeader header, RsaPubKey rsa, byte[] pubKeyData)
+        {
+            var snPubKey = new SnPublicKeyBlob()
+            {
+                SigAlgId = AlgorithmId.RsaSign,
+                HashAlgId = AlgorithmId.Sha,
+                PublicKeySize = (UInt32)(s_offsetToKeyData + pubKeyData.Length)
+            };
+
+            using (var ms = new MemoryStream(160))
+            using (var binaryWriter = new BinaryWriter(ms))
+            {
+                binaryWriter.Write(snPubKey.SigAlgId);
+                binaryWriter.Write(snPubKey.HashAlgId);
+                binaryWriter.Write(snPubKey.PublicKeySize);
+
+                binaryWriter.Write(header.Type);
+                binaryWriter.Write(header.Version);
+                binaryWriter.Write(header.Reserved);
+                binaryWriter.Write(header.AlgId);
+
+                binaryWriter.Write(rsa.Magic);
+                binaryWriter.Write(rsa.BitLen);
+                binaryWriter.Write(rsa.PubExp);
+
+                binaryWriter.Write(pubKeyData);
+
+                return ms.ToArray();
+            }
+        }
+
+        internal unsafe static bool TryGetPublicKeyFromPrivateKeyBlob(byte[] blob, out byte[] publicKey)
+        {
+            fixed (byte* blobPtr = blob)
+            {
+                var header = (BlobHeader*)blobPtr;
+                var rsa = (RsaPubKey*)(blobPtr + sizeof(BlobHeader));
+
+                var version = header->Version;
+                var modulusBitLength = rsa->BitLen;
+                var exponent = rsa->PubExp;
+                var modulus = new byte[modulusBitLength >> 3];
+
+                if (blob.Length - s_offsetToKeyData < modulus.Length)
+                {
+                    publicKey = null;
+                    return false;
+                }
+
+                Marshal.Copy((IntPtr)(blobPtr + s_offsetToKeyData), modulus, 0, modulus.Length);
+
+                var newHeader = new BlobHeader()
+                {
+                    Type = PublicKeyBlobId,
+                    Version = version,
+                    Reserved = 0,
+                    AlgId = AlgorithmId.RsaSign
+                };
+
+                var newRsaKey = new RsaPubKey()
+                {
+                    Magic = RSA1, // Public key
+                    BitLen = modulusBitLength,
+                    PubExp = exponent
+                };
+
+                publicKey = CreateSnPublicKeyBlob(newHeader, newRsaKey, modulus);
+                return true;
+            }
+        }
+
+        private struct AlgorithmId
+        {
+            // From wincrypt.h
+            private const int AlgorithmClassOffset = 13;
+            private const int AlgorithmClassMask = 0x7;
+            private const int AlgorithmSubIdOffset = 0;
+            private const int AlgorithmSubIdMask = 0x1ff;
+
+            private readonly uint _flags;
+
+            public const int RsaSign = 0x00002400;
+            public const int Sha = 0x00008004;
+
+            public bool IsSet
+            {
+                get { return _flags != 0; }
+            }
+
+            public AlgorithmId(uint flags)
+            {
+                _flags = flags;
+            }
+        }
+    }
+}

--- a/src/Nerdbank.GitVersioning.Tasks/Nerdbank.GitVersioning.Tasks.csproj
+++ b/src/Nerdbank.GitVersioning.Tasks/Nerdbank.GitVersioning.Tasks.csproj
@@ -2,13 +2,11 @@
   <PropertyGroup>
     <TargetFrameworks>netcoreapp1.0;net45</TargetFrameworks>
     <IsTool>true</IsTool>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
   </PropertyGroup>
   <ItemGroup>
     <PackageReference Include="Nerdbank.GitVersioning.LKG" Version="1.6.20-beta-gfea83a8c9e" />
     <PackageReference Include="NuProj.Common" Version="0.11.14-beta" />
-  </ItemGroup>
-  <ItemGroup Condition="'$(TargetFramework)' == 'net45' ">
-    <PackageReference Include="PInvoke.MSCorEE" Version="0.3.152" />
   </ItemGroup>
   <ItemGroup>
     <ProjectReference Include="..\MSBuildExtensionTask\MSBuildExtensionTask.csproj" />

